RadixWallet/Prelude/DefaultValueDecodable.swift

```diff
@@ -0,0 +1,82 @@
+import Foundation
+import IdentifiedCollections
+
+// MARK: - DefaultValueProvider
+/// An utility property wrapper allowing to specify a default value if the decoded value is missing or nil
+
+/// Provides the Type of the decoded value, also its default value
+public protocol DefaultValueProvider {
+	associatedtype Value: Codable
+	static var defaultValue: Value { get }
+}
+
+// MARK: - DefaultCodable
+public enum DefaultCodable {
+	/// The property wrapper itself.
+	/// Need to make use of the tagging with the protocol, since the value provided in `init` will not be
```

Comment:
I'm pretty sure I've written DefaultCodable in the simple straightforward way, are you sure it's not possible to just provide the value as an argument?
```
@DefaultCodable([])
var flags: Flags
```

Reply:
Or maybe not actually, seems like I never manage to solve that at the property level, instead I made a type level protocol `DefaultingType` that in this case `Flags` would conform to. Not necessarily better than your solution, but I'll paste it here for comparison:

```

import Foundation

// MARK: - DefaultingType

/**
 `DefaultingType` is a protocol that makes sure that a type, when used in a property, is never `nil`, even if the value is missing when decoding.
 Its requirement is that the type has a static var `defaultValue` of type `Self`. For enums, this can also be fulfilled by having an enum case called `defaultValue`.

     struct Location: DefaultingType {
         let lat: Int
         let lon: Int

         static let defaultValue: Self = .init(lat: 0, lon: 0)
     }

     struct User: Decodable {
         // Will never be decoded to nil
         let location: Location

         // Any corrupted values will be decoded to defaultValue
         let locations: [Location]
     }
 */
public protocol DefaultingType: Decodable {
  static var defaultValue: Self { get }
}

/// A helper type to decode `DefaultingType` values
private struct DefaultingWrapper<T: DefaultingType>: Decodable {
  let value: T

  public init(from decoder: Decoder) throws {
    self.value = (try? .init(from: decoder)) ?? .defaultValue
  }
}

extension KeyedDecodingContainer {
  public func decode<T: DefaultingType>(_ type: T.Type, forKey key: Self.Key) throws -> T {
    (try? decode(DefaultingWrapper.self, forKey: key).value) ?? .defaultValue
  }

  public func decodeIfPresent<T: DefaultingType>(_ type: T.Type, forKey key: Self.Key) throws -> T? {
    try decodeIfPresent(DefaultingWrapper.self, forKey: key)?.value
  }

  public func decode<T: DefaultingType>(_ type: [T].Type, forKey key: Self.Key) throws -> [T] {
    guard let result = try? decode([DefaultingWrapper<T>].self, forKey: key) else { return [] }
    return result.map(\.value)
  }

  public func decodeIfPresent<T: DefaultingType>(_ type: [T].Type, forKey key: Self.Key) throws -> [T]? {
    guard let result = try? decodeIfPresent([DefaultingWrapper<T>].self, forKey: key) else { return nil }
    return result.map(\.value)
  }
}
```

Reply:
I changed \`\`\` into \`\`\`swift" for Swift colour highlighting

Reply:
@kugel3 Problem with being able to do just:
```
@DefaultCodable([])
var flags: Flags
```

Is that the given default value will not be used when initialised with Decoder, and it makes sense, since it is a different initialiser. So we have to use this whole dance with protocols.

Reply:
Yeah I remembered that when I looked at my old code, that's why I also had to resort to type level solutions. But there are discussions on Swift evolution that might fix this